### ISSUE TITLE
gevent.local.local() memory leak

### ISSUE THREAD

**Issue by sublee** [gevent/gevent#981]
* gevent version: 1.2.2
* Python version: CPython 2.7.12, PyPy 5.8.0-2.7.13
* Operating System: Ubuntu 16.04

### Description:

There's a memory leak when making `gevent.local.local()` In a greenlet.

I digged a bit.  The `clear` callback in [gevent/local.py#L208](https://github.com/gevent/gevent/blob/1.2.2/src/gevent/local.py#L208) looks like the leak point.  Perhaps this issue appears from 1192b1c6dc302d346b2388a57297b201df236973.

### What I've run:

You can reproduce the memory leak by the below simple code:

```python
import gc
import gevent.local
def main():
    while True:
        gevent.local.local()
        gc.collect()
gevent.spawn(main).join()
```

The growth is much faster on PyPy.

**Comment by jamadden** [comment]
It is by design that the local dictionaries aren't cleared until the greenlet exits, so I don't believe that's a leak. If you allocate and never destroy the greenlet they will not be cleared. I believe the prior implementation had the same behaviour (in fact it may have been even less timely).

Most locals are allocated at a static scope (e.g., module or class level) and assigned to a variable from which they are reachable, or occasionally they are allocated as instance attributes, and again, assigned to a variable. They have to be named to be useful, so simply allocating them in a loop is not a good representation of a real scenario.

**Comment by sublee** [comment]
As you said, we can make a local for an instance attribute.  At this case, if we have some infinite greenlets that have touched the local at least once, the callbacks from the local will never be released even though the instance is deleted.

I just wanted to show you the memory leak from deleted locals.  So I put the allocation without name in the loop.  The below example would be more practical:

```python
import gc
import gevent.local

class O(object):
    def __init__(self):
        self.local = gevent.local.local()

def main_loop():
    x = 0
    objs = []
    while True:
        o = O()
        objs.append(o)
        x += 1
        if x % 1000 == 0:
            del objs[:]
            gc.collect()

gevent.spawn(main_loop).join()
```

**Comment by sublee** [comment]
Awesome!  Thank you so much!

**Comment by hellobiek** [comment]
Has this problem been solved?

**Comment by jamadden** [comment]
As can be seen above, this issue was closed as fixed with the merge of #983
